## Goals: 

Melakukan Prediksi 'Sale Price' dari data yang diberikan 

In [1]:
## Uncomment apabila mau install pacakges
# %pip install numpy pandas seaborn matplotlib scikit-learn

#### Import Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


# Configurasi pandas untuk display semua kolom
pd.set_option('display.max_columns', None)

/var/folders/my/rbl5q7yd7pxb2yk9_qjc7mjr0000gn/T/ipykernel_12000/553640303.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


#### Load File 

load file train & test

In [3]:
file = pd.read_csv('./train.csv')
dfOriginal = pd.DataFrame(file).drop(columns=('Id'))

fileTest = pd.read_csv('./test.csv')
dfTest = pd.DataFrame(fileTest).drop(columns=('Id'))

#### Gabungkan file dan berikan index

In [4]:
df = result_df = pd.concat([dfOriginal, dfTest], ignore_index=True)

INDEXES = {
  'train': {
    'start': 0,
    'end': dfOriginal.shape[0] - 1,
  },
  'test': {
    'start': dfOriginal.shape[0],
    'end': dfOriginal.shape[0] - 1 + dfTest.shape[0] - 1
  }
}

df

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0             60       RL         65.0     8450   Pave   NaN      Reg   
1             20       RL         80.0     9600   Pave   NaN      Reg   
2             60       RL         68.0    11250   Pave   NaN      IR1   
3             70       RL         60.0     9550   Pave   NaN      IR1   
4             60       RL         84.0    14260   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
2914         160       RM         21.0     1936   Pave   NaN      Reg   
2915         160       RM         21.0     1894   Pave   NaN      Reg   
2916          20       RL        160.0    20000   Pave   NaN      Reg   
2917          85       RL         62.0    10441   Pave   NaN      Reg   
2918          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
2914         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
2915         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
2916         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
2917         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
2918         Lvl    AllPub    Inside       Mod      Mitchel       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
2914       Norm    Twnhs     2Story            4            7       1970   
2915       Norm   TwnhsE     2Story            4            5       1970   
2916       Norm     1Fam     1Story            5            7       1960   
2917       Norm     1Fam     SFoyer            5            5       1992   
2918       Norm     1Fam     2Story            7            5       1993   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
2914          1970     Gable  CompShg     CemntBd     CmentBd        NaN   
2915          1970     Gable  CompShg     CemntBd     CmentBd        NaN   
2916          1996     Gable  CompShg     VinylSd     VinylSd        NaN   
2917          1992     Gable  CompShg     HdBoard     Wd Shng        NaN   
2918          1994     Gable  CompShg     HdBoard     HdBoard    BrkFace   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   


#### Mengecek head & tail
memastikan bahwa file yang kita import sudah benar

In [5]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ       706.0          Unf         0.0      150.0        856.0   
1          ALQ       978.0          Unf         0.0      284.0       1262.0   
2          GLQ       486.0          Unf         0.0      434.0        920.0   
3          ALQ       216.0          Unf         0.0      540.0        756.0   
4          GLQ       655.0          Unf         0.0      490.0       1145.0   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710           1.0           0.0         2         1             3   
1       1262           0.0           1.0         2         0             3   
2       1786           1.0           0.0         2         1             3   
3       1717           1.0           0.0         1         0             3   
4       2198           1.0           0.0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd             7        Typ           

In [6]:
df.tail()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
2914         160       RM         21.0     1936   Pave   NaN      Reg   
2915         160       RM         21.0     1894   Pave   NaN      Reg   
2916          20       RL        160.0    20000   Pave   NaN      Reg   
2917          85       RL         62.0    10441   Pave   NaN      Reg   
2918          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
2914         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
2915         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
2916         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
2917         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
2918         Lvl    AllPub    Inside       Mod      Mitchel       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
2914       Norm    Twnhs     2Story            4            7       1970   
2915       Norm   TwnhsE     2Story            4            5       1970   
2916       Norm     1Fam     1Story            5            7       1960   
2917       Norm     1Fam     SFoyer            5            5       1992   
2918       Norm     1Fam     2Story            7            5       1993   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
2914          1970     Gable  CompShg     CemntBd     CmentBd        NaN   
2915          1970     Gable  CompShg     CemntBd     CmentBd        NaN   
2916          1996     Gable  CompShg     VinylSd     VinylSd        NaN   
2917          1992     Gable  CompShg     HdBoard     Wd Shng        NaN   
2918          1994     Gable  CompShg     HdBoard     HdBoard    BrkFace   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
2914         0.0        TA        TA     CBlock       TA       TA   
2915         0.0        TA        TA     CBlock       TA       TA   
2916         0.0        TA        TA     CBlock       TA       TA   
2917         0.0        TA        TA      PConc       Gd       TA   
2918        94.0        TA        TA      PConc       Gd       TA   

     BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
2914           No          Unf         0.0          Unf         0.0   
2915           No          Rec       252.0          Unf         0.0   
2916           No          ALQ      1224.0          Unf         0.0   
2917           Av          GLQ       337.0          Unf         0.0   
2918           Av          LwQ       758.0          Unf         0.0   

      BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
2914      546.0        546.0    GasA        Gd          Y      SBrkr   
2915      294.0        546.0    GasA        TA          Y      SBrkr   
2916        0.0       1224.0    GasA        Ex          Y      SBrkr   
2917      575.0        912.0    GasA        TA          Y      SBrkr   
2918      238.0        996.0    GasA        Ex          Y      SBrkr   

      1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
2914       546       546             0       1092           0.0           0.0   
2915       546       546             0       1092           0.0           0.0   
2916      1224         0             0       1224           1.0           0.0   
2917       970         0             0        970           0.0           1.0   
2918       996      1004             0       2000           0.0           0.0   

      FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  \
2914         1         1             3             1          TA   
2915         1         1             3             1          TA   
2916         1         0             4             1          TA   
2917         1         0             3             1          TA   
2918         2         1             3             1          TA   

      TotRmsAbvGrd Functional  Firepl

#### mengambil informasi kolom yang bersifat Kategorikal & Numerikal

In [7]:
categoryTypeColumns = list(df.select_dtypes(exclude=['number']).columns)
numericalColumns = list(df.select_dtypes(include=['int64', 'float64']))

#### Memproses data yang kosong 

Setiap kolom kategorikal, diisi value yang `null` dengan nilai yang sering muncul pada kolom tersebut dengan method `fillna`


In [8]:
# Fill null or N/A to categorical columns
df[categoryTypeColumns] = df[categoryTypeColumns].apply(lambda col: col.fillna(col.mode()[0]))

#### Deklarasi functions yang diperlukan 

In [9]:
def processCategoricalColumnsWithOneHotEncoding(dataFrame, categoryTypeColumns):
  # convert categorical to true false
  for var in categoryTypeColumns:
    cat_list = pd.get_dummies(dataFrame[var], prefix=var)
    dataFrame = dataFrame.join(cat_list)

  # remove categorical
  allColumns = dataFrame.columns.values.tolist()
  keepColumns = [i for i in allColumns if i not in categoryTypeColumns]
  dataFrame = dataFrame[keepColumns]

  return dataFrame

#### Konversi kolom kategorikal dengan one-hot-encoding
kolom yang bersifat kategorikal menjadi nilai absolut yang bisa dibaca oleh komputer dengan metode one-hot-encoding


In [10]:
# convert categorical columns with one-hot-encoding
dfConverted = processCategoricalColumnsWithOneHotEncoding(dataFrame=df, categoryTypeColumns=categoryTypeColumns)

#### Pencekan Head & Tail file
memastikan bahwa dataframe berhasil di-convert

In [11]:
dfConverted.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0          60         65.0     8450            7            5       2003   
1          20         80.0     9600            6            8       1976   
2          60         68.0    11250            7            5       2001   
3          70         60.0     9550            7            5       1915   
4          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          2003       196.0       706.0         0.0      150.0        856.0   
1          1976         0.0       978.0         0.0      284.0       1262.0   
2          2002       162.0       486.0         0.0      434.0        920.0   
3          1970         0.0       216.0         0.0      540.0        756.0   
4          2000       350.0       655.0         0.0      490.0       1145.0   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0       856       854             0       1710           1.0           0.0   
1      1262         0             0       1262           0.0           1.0   
2       920       866             0       1786           1.0           0.0   
3       961       756             0       1717           1.0           0.0   
4      1145      1053             0       2198           1.0           0.0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0         2         1             3             1             8           0   
1         2         0             3             1             6           1   
2         2         1             3             1             6           1   
3         1         0             3             1             7           1   
4         2         1             4             1             9           1   

   GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
0       2003.0         2.0       548.0           0           61   
1       1976.0         2.0       460.0         298            0   
2       2001.0         2.0       608.0           0           42   
3       1998.0         3.0       642.0           0           35   
4       2000.0         3.0       836.0         192           84   

   EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  \
0              0          0            0         0        0       2    2008   
1              0          0            0         0        0       5    2007   
2              0          0            0         0        0       9    2008   
3            272          0            0         0        0       2    2006   
4              0          0            0         0        0      12    2008   

   SalePrice  MSZoning_C (all)  MSZoning_FV  MSZoning_RH  MSZoning_RL  \
0   208500.0             False        False        False         True   
1   181500.0             False        False        False         True   
2   223500.0             False        False        False         True   
3   140000.0             False        False        False         True   
4   250000.0             False        False        False         True   

   MSZoning_RM  Street_Grvl  Street_Pave  Alley_Grvl  Alley_Pave  \
0        False        False         True        True       False   
1        False        False         True        True       False   
2        False        False         True        True       False   
3        False        False         True        True       False   
4        False        False         True        True       False   

   LotShape_IR1  LotShape_IR2  LotShape_IR3  LotShape_Reg  LandContour_Bnk  \
0         False         False         False          True            False   
1         False         False         False          True            False   
2          True         False         False         False            False   
3          True         False         False         False            False   
4          

In [12]:
dfConverted.tail()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
2914         160         21.0     1936            4            7       1970   
2915         160         21.0     1894            4            5       1970   
2916          20        160.0    20000            5            7       1960   
2917          85         62.0    10441            5            5       1992   
2918          60         74.0     9627            7            5       1993   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
2914          1970         0.0         0.0         0.0      546.0   
2915          1970         0.0       252.0         0.0      294.0   
2916          1996         0.0      1224.0         0.0        0.0   
2917          1992         0.0       337.0         0.0      575.0   
2918          1994        94.0       758.0         0.0      238.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
2914        546.0       546       546             0       1092           0.0   
2915        546.0       546       546             0       1092           0.0   
2916       1224.0      1224         0             0       1224           1.0   
2917        912.0       970         0             0        970           0.0   
2918        996.0       996      1004             0       2000           0.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
2914           0.0         1         1             3             1   
2915           0.0         1         1             3             1   
2916           0.0         1         0             4             1   
2917           1.0         1         0             3             1   
2918           0.0         2         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
2914             5           0          NaN         0.0         0.0   
2915             6           0       1970.0         1.0       286.0   
2916             7           1       1960.0         2.0       576.0   
2917             6           0          NaN         0.0         0.0   
2918             9           1       1993.0         3.0       650.0   

      WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  \
2914           0            0              0          0            0   
2915           0           24              0          0            0   
2916         474            0              0          0            0   
2917          80           32              0          0            0   
2918         190           48              0          0            0   

      PoolArea  MiscVal  MoSold  YrSold  SalePrice  MSZoning_C (all)  \
2914         0        0       6    2006        NaN             False   
2915         0        0       4    2006        NaN             False   
2916         0        0       9    2006        NaN             False   
2917         0      700       7    2006        NaN             False   
2918         0        0      11    2006        NaN             False   

      MSZoning_FV  MSZoning_RH  MSZoning_RL  MSZoning_RM  Street_Grvl  \
2914        False        False        False         True        False   
2915        False        False        False         True        False   
2916        False        False         True        False        False   
2917        False        False         True        False        False   
2918        False        False         True        False        False   

      Street_Pave  Alley_Grvl  Alley_Pave  LotShape_IR1  LotShape_IR2  \
2914         True        True       False         False         False   
2915         True        True       False         False         False   
2916         True        True       False         False         False   
2917         True        True       False         False         False   
2918         True        True       False         False         False   

      LotShape_IR3  LotShape_Reg  LandConto

#### Memproses data yang kosong 

Setiap kolom numerikal, diisi value yang `null` dengan nilai rata-rata pada kolom tersebut dengan method `fillna`


In [13]:
# fill out N/A value with mean value
dfConverted.fillna(dfConverted.mean(numeric_only=True), inplace=True)

#### Memastikan bahwa tidak ada nilai `null` pada dataframe

In [14]:
missingData = dfConverted.isnull().sum()
dataTypeColumns = dfConverted.dtypes
dataTypeColumns
pd.set_option('display.max_rows', None)

missingDataDf = pd.DataFrame({'Missing Value': missingData, 'Data Type': dataTypeColumns})

print(missingDataDf)
pd.reset_option('display.max_rows')

# Expet missing value / null value is 0

                       Missing Value Data Type
MSSubClass                         0     int64
LotFrontage                        0   float64
LotArea                            0     int64
OverallQual                        0     int64
OverallCond                        0     int64
YearBuilt                          0     int64
YearRemodAdd                       0     int64
MasVnrArea                         0   float64
BsmtFinSF1                         0   float64
BsmtFinSF2                         0   float64
BsmtUnfSF                          0   float64
TotalBsmtSF                        0   float64
1stFlrSF                           0     int64
2ndFlrSF                           0     int64
LowQualFinSF                       0     int64
GrLivArea                          0     int64
BsmtFullBath                       0   float64
BsmtHalfBath                       0   float64
FullBath                           0     int64
HalfBath                           0     int64
BedroomAbvGr 

#### Pisahkan dataframe 
Train & Test yang telah digabungkan dipisahkan berdasarkan nomor index yang sudah didefinisikan diawal

In [15]:
# separate Train and Test DataFrames
dfTrain = dfConverted.loc[:INDEXES['train']['end']]
dfTest = dfConverted.loc[INDEXES['test']['start']:INDEXES['test']['end']]

# get X and Y train
X_train = dfTrain.drop(columns=('SalePrice'), axis=1)
y_train = dfTrain['SalePrice']

# get X Test
X_test = dfTest.drop(columns=('SalePrice'), axis=1)

dfTrain

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0             60         65.0     8450            7            5       2003   
1             20         80.0     9600            6            8       1976   
2             60         68.0    11250            7            5       2001   
3             70         60.0     9550            7            5       1915   
4             60         84.0    14260            8            5       2000   
...          ...          ...      ...          ...          ...        ...   
1455          60         62.0     7917            6            5       1999   
1456          20         85.0    13175            6            6       1978   
1457          70         66.0     9042            7            9       1941   
1458          20         68.0     9717            5            6       1950   
1459          20         75.0     9937            5            6       1965   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0             2003       196.0       706.0         0.0      150.0   
1             1976         0.0       978.0         0.0      284.0   
2             2002       162.0       486.0         0.0      434.0   
3             1970         0.0       216.0         0.0      540.0   
4             2000       350.0       655.0         0.0      490.0   
...            ...         ...         ...         ...        ...   
1455          2000         0.0         0.0         0.0      953.0   
1456          1988       119.0       790.0       163.0      589.0   
1457          2006         0.0       275.0         0.0      877.0   
1458          1996         0.0        49.0      1029.0        0.0   
1459          1965         0.0       830.0       290.0      136.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0           856.0       856       854             0       1710           1.0   
1          1262.0      1262         0             0       1262           0.0   
2           920.0       920       866             0       1786           1.0   
3           756.0       961       756             0       1717           1.0   
4          1145.0      1145      1053             0       2198           1.0   
...           ...       ...       ...           ...        ...           ...   
1455        953.0       953       694             0       1647           0.0   
1456       1542.0      2073         0             0       2073           1.0   
1457       1152.0      1188      1152             0       2340           0.0   
1458       1078.0      1078         0             0       1078           1.0   
1459       1256.0      1256         0             0       1256           1.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0              0.0         2         1             3             1   
1              1.0         2         0             3             1   
2              0.0         2         1             3             1   
3              0.0         1         0             3             1   
4              0.0         2         1             4             1   
...            ...       ...       ...           ...           ...   
1455           0.0         2         1             3             1   
1456           0.0         2         0             3             1   
1457           0.0         2         0             4             1   
1458           0.0         1         0             2             1   
1459           0.0         1         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0                8           0       2003.0         2.0       548.0   
1                6           1       1976.0         2.0       460.0   
2                6           1       2001.0         2.0       608.0   
3                7           1       1998.0         3.0       642.0   
4                9           1       2000.0         3.0       836.0   


#### Modeling dengan Random Forest Regressor

In [16]:
# Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor

modelRFRegressor = RandomForestRegressor(n_estimators=1000, random_state=42)
modelRFRegressor.fit(X_train, y_train)


RandomForestRegressor(n_estimators=1000, random_state=42)

#### Mengecek score model dengan method `.score()`

In [17]:
scoreRFRegressor = modelRFRegressor.score(X_train,y_train)
print(f'Score: {scoreRFRegressor}')

predictedSalePricesRFR = modelRFRegressor.predict(X_test)
print(predictedSalePricesRFR)

Score: 0.9814536638770964
[128665.623 154804.272 181979.923 ...  86482.998 152694.299 114888.554]


bisa disimpulkan score yang dihasilkan dengan **Random Forest Regressor** 

n_estimators = `1000`

random_state = `42`

mendapatkan score `0.98`


---

#### Modeling dengan Linear Regression

In [18]:
# Linear Regression
from sklearn.linear_model import LinearRegression

modelLinearRegression = LinearRegression()
modelLinearRegression.fit(X_train, y_train)

LinearRegression()

#### Mengecek score model dengan method `.score()`

In [19]:
scoreModelLinearRegression = modelLinearRegression.score(X_train, y_train)
print(f'Score: {scoreModelLinearRegression}')

predictedSalePricesLR = modelLinearRegression.predict(X_test)
print(predictedSalePricesLR)

Score: 0.9321714155795846
[112961.28218504 154285.05642433 186818.12510115 ...  79238.13613556
 179620.32420002 107898.98551219]


bisa disimpulkan score yang dihasilkan dengan **Linear Regression** 

mendapatkan score `0.93`

lebih rendah dibandingkan dengan menggunakan **RFR**


---

Dilihat dari scorenya maka model yang lebih baik digunakan yaitu Random Forest Regressor

---
